rust/sbp/src/parser.rs

```diff
@@ -0,0 +1,366 @@
+    }
+    Ok(v)
+}
+
+pub fn read_s16_array_limit(buf: &mut &[u8], n: usize) -> Result<Vec<i16>> {
+    let mut v = Vec::new();
+    for _ in 0..n {
+        v.push(buf.read_i16::<LittleEndian>()?);
+    }
+    Ok(v)
+}
+
+pub fn read_u16_array_limit(buf: &mut &[u8], n: usize) -> Result<Vec<u16>> {
+    let mut v = Vec::new();
+    for _ in 0..n {
+        v.push(buf.read_u16::<LittleEndian>()?);
+    }
+    Ok(v)
+}
+
+pub fn read_float_array_limit(buf: &mut &[u8], n: usize) -> Result<Vec<f32>> {
+    let mut v = Vec::new();
+    for _ in 0..n {
+        v.push(buf.read_f32::<LittleEndian>()?);
+    }
+    Ok(v)
+}
+
+pub fn read_double_array_limit(buf: &mut &[u8], n: usize) -> Result<Vec<f64>> {
+    let mut v = Vec::new();
+    for _ in 0..n {
+        v.push(buf.read_f64::<LittleEndian>()?);
+    }
+    Ok(v)
+}
+
+#[derive(Debug)]
+pub enum ParseResult {
+    Ok((usize, SBP)),
+    Err((usize, Error)),
+    Incomplete,
+}
+
+pub fn parse_sbp<'a>(input: &'a [u8]) -> ParseResult {
+    let (new_input, mut frame) = match frame::parse(input) {
+        Ok(res) => res,
+        Err(e) => match e {
+            NomErr::Incomplete(_) => return ParseResult::Incomplete,
+            NomErr::Error((_, kind)) | NomErr::Failure((_, kind)) => {
+                return ParseResult::Err((1, Error::ParseError { kind }))
+            }
+        },
+    };
+
+    let bytes_read = input.len() - new_input.len();
+
+    if !frame::check_crc(&frame) {
+        return ParseResult::Err((
+            bytes_read,
+            Error::CrcError {
+                msg_type: frame.1,
+                sender_id: frame.2,
+                crc: frame.4,
+            },
+        ));
+    }
+
+    let msg = match SBP::parse(frame.1, frame.2, &mut frame.3) {
+        Ok(msg) => msg,
+        Err(err) => return ParseResult::Err((bytes_read, err)),
+    };
+
+    ParseResult::Ok((bytes_read, msg))
+}
+
+pub mod frame {
+    use nom::{bytes, error::ErrorKind, number, sequence::tuple, IResult};
+
+    pub type Frame<'a> = (&'a [u8], u16, u16, &'a [u8], u16);
+
+    pub type Error<'a> = (&'a [u8], ErrorKind);
+
+    pub fn parse(i: &[u8]) -> IResult<&[u8], Frame, Error> {
+        let (i, _) = bytes::streaming::take_while(is_not_preamble)(i)?;
+
+        tuple((
+            parse_preamble,
+            parse_msg_type,
+            parse_sender_id,
+            parse_payload,
+            parse_crc,
+        ))(i)
+    }
+
+    pub fn check_crc((_preamble, msg_type, sender_id, payload, crc_in): &Frame) -> bool {
+        let mut crc = crc16::State::<crc16::XMODEM>::new();
+
+        crc.update(&msg_type.to_le_bytes());
+        crc.update(&sender_id.to_le_bytes());
+        crc.update(&[payload.len() as u8]);
+        crc.update(payload);
+
+        crc.get() == *crc_in
+    }
+
+    fn is_not_preamble(b: u8) -> bool {
+        b != 0x55
+    }
+
+    fn parse_preamble(i: &[u8]) -> IResult<&[u8], &[u8], Error> {
+        bytes::streaming::is_a("\x55")(i)
```

Comment:
Note: code here is mostly the same with the exception that we're not using the streaming interface from nom, this adds "incomplete" as a possible return value, which integrates better with the tokio codec code